lib/auth/u2f/authenticate.go

```diff
@@ -0,0 +1,207 @@
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err = params.Storage.UpsertU2FSignChallenge(params.StorageKey, challenge); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return challenge.SignRequest(*reg), nil
+}
+
+// AuthenticateSignChallenge is the second step in the authentication sequence.
+// It runs on the client and the returned AuthenticationChallengeResponse must
+// be sent to the server.
+//
+// Note: this function writes user interaction prompts to stdout.
+func AuthenticateSignChallenge(c AuthenticateChallenge, facet string) (*AuthenticateChallengeResponse, error) {
+	// Convert request struct between the different libraries.
+	req := &u2fhost.AuthenticateRequest{
+		Challenge: c.Challenge,
+		AppId:     c.AppID,
+		Facet:     facet,
+		KeyHandle: c.KeyHandle,
+		CheckOnly: true,
+	}
+
+	// Open available U2F devices.
+	allDevices := u2fhost.Devices()
+	if len(allDevices) == 0 {
+		return nil, trace.NotFound("no u2f devices found, please plug one in to authenticate")
+	}
+	// Filter only the devices that can be opened and are registered for this user.
+	var openDevices []u2fhost.Device
+	for _, device := range allDevices {
+		if err := device.Open(); err != nil {
+			continue
+		}
+		// Call Authenticate with CheckOnly set in AuthenticateRequest, which
+		// will check whether this device has the provided KeyHandle.
+		if _, err := device.Authenticate(req); err != nil {
+			// BadKeyHandleError is expected when this device doesn't have the
+			// KeyHandle.
+			if _, ok := err.(*u2fhost.BadKeyHandleError); ok {
+				continue
+			}
+			// TestOfUserPresenceRequiredError is expected when this device
+			// *does* have the KeyHandle. Any other kind of error is
+			// unexpected.
+			if _, ok := err.(*u2fhost.TestOfUserPresenceRequiredError); !ok {
+				continue
+			}
+		}
+
+		openDevices = append(openDevices, device)
+		defer func(device u2fhost.Device) {
+			device.Close()
+		}(device)
+	}
+	if len(openDevices) == 0 {
+		return nil, trace.NotFound("found %d u2f devices, but none of them are registered with this Teleport user", len(allDevices))
+	}
+
+	fmt.Println("Please press the button on your U2F key")
+	// Unset CheckOnly to perform the actual challenge signing.
+	req.CheckOnly = false
+
+	var res *u2fhost.AuthenticateResponse
+	var err error
+	start := time.Now()
+outer:
+	for {
+		if time.Since(start) > time.Minute {
+			return nil, trace.LimitExceeded("timed out waiting for a U2F key press")
+		}
+		// The below device.Authenticate calls are non-blocking. They return
+		// u2fhost.TestOfUserPresenceRequiredError until the user actually
+		// touches the device button.
+		for _, device := range openDevices {
+			res, err = device.Authenticate(req)
+			if err != nil {
+				if _, ok := err.(*u2fhost.TestOfUserPresenceRequiredError); ok {
+					continue
+				}
+				return nil, trace.Wrap(err)
+			}
+			break outer
+		}
+		time.Sleep(200 * time.Millisecond)
+	}
+
```

Comment:
I would consider checking `res` for `nil` upon the exit from the outer loop, just in case.